build-monitor-plugin/src/main/java/com/smartcodeltd/jenkinsci/plugins/buildmonitor/viewmodel/BuildView.java

```diff
@@ -144,36 +145,37 @@ public BuildViewModel previousBuild() {
 
     @Override
     public Set<String> culprits() {
-        return getUsers(new Reader() {
-            @Override
-            public Iterable<String> readUsersFrom(AbstractBuild<?, ?> jenkinsBuild) {
-                return transform(jenkinsBuild.getCulprits(), new Function<User, String>() {
-                    @Override
-                    public String apply(User culprit) {
-                        return culprit.getFullName();
-                    }
-                });
+        Set<String> culprits = new TreeSet<String>();
+        if (build instanceof AbstractBuild<?, ?>) {
+            Iterable<String> forAbstractBuild = transform(((AbstractBuild<?, ?>) build).getCulprits(), new Function<User, String>() {
+                @Override
+                public String apply(User culprit) {
+                    return culprit.getFullName();
+                }
+            });
+            Iterables.addAll(culprits, forAbstractBuild);
+        } else if (build instanceof WorkflowRun) {
+            //Workaround while waiting for https://issues.jenkins-ci.org/browse/JENKINS-24141
+            WorkflowRun previous = ((WorkflowRun) build).getPreviousCompletedBuild();
+            if (build.isBuilding()) {
+                //We are currently building so add culprits from previous build (if any)
+                if (previous != null && previous.getResult().isWorseThan(Result.SUCCESS)) {
+                    culprits.addAll(getCommitters(previous));
+                }
+            }
+            culprits.addAll(getCommitters(build));
+
+            //Get culprits from earlier builds
+            if (previous != null && previous.getPreviousNotFailedBuild() != null) {
+                culprits.addAll(getCulpritsForWorkflowRun(previous.getPreviousNotFailedBuild(), previous));
             }
-        });
+        }
+        return culprits;
     }
 
     @Override
     public Set<String> committers() {
-        return getUsers(new Reader() {
-            @Override
-            public Iterable<String> readUsersFrom(AbstractBuild<?, ?> jenkinsBuild) {
-                return transform(nonNullIterable(jenkinsBuild.getChangeSet()), new Function<ChangeLogSet.Entry, String>() {
-                    @Override
-                    public String apply(ChangeLogSet.Entry entry) {
-                        return entry.getAuthor().getFullName();
-                    }
-                });
-            }
-
-            private <T> T nonNullIterable(T list) {
-                return (T) getOrElse(list, newArrayList());
-            }
-        });
+        return getCommitters(build);
```

Comment:
Could we extract this logic into a separate class? I'm worried that the BuildView is doing quite a few things already.